### ISSUE TITLE
add ability to run more than one consul release per cluster

### ISSUE THREAD

**Issue by iameli** [hashicorp/consul-helm#803]
#### Is your feature request related to a problem? Please describe.

Installing this chart multiple times on the same cluster results in the second deployment having lots and lots of hostPort conflicts.

#### Feature Description

All of the hostPorts should be customizable, so you can run more than one Consul per Kubernetes cluster if you care to.

#### Use Case(s)

I want to run a "staging Consul" and a "production Consul" on the same nodes.

#### Contributions

Yep, I've already started — will send a PR with my preliminary work once I get something running.

Mostly filing this in case there's a way to do this that I'm not seeing.

**Comment by ndhanushkodi** [comment]
Hey @iameli, this seems similar to #126 and #597. So in addition to the hostPort conflicts in consul-helm, there's also lots of automation that's done in consul-k8s that requires specific ports and doesn't support multiple installations on the same cluster. I am not sure if this is something we'd want to support in the long term, since most production clusters would ideally have just one consul installation. We do really appreciate you wanting to contribute the change, but can we see if there's more feedback that this is a use case that a lot of people need before deciding to support it in the long term?

**Comment by iameli** [comment]
Got it. Let me explain a a little about our use case, perhaps there's a smarter way to accomplish what I'm trying to do. Livepeer is building a global network of high-availability RTMP servers to provide video ingest into our decentralized transcoding network. Toward that end, we've bought or leased bare-metal servers all over the world, which we've tied together in to Kubernetes clusters with k3s.

How we're currently using Consul: when livestreams come in, the RTMP server that handles the incoming TCP connection writes an entry to Consul KV. Incoming HTTPS connections hit Traefik, and Traefik uses that data to route HTTPS playback to the appropriate origin server presently handling that connection. This is how we provide load distribution and origin discovery, and it's working great.

How we'd like to use Consul: If I federate all of our currently-independent Consul servers, and muck around with consul-replicate, I can push that same KV entry into all of our datacenters. This will allow us to have a single playback URL, `cdn.livepeer.com`, that can route traffic appropriately from each of the datacenters. Neat! But, two problems:

1. I can't justify buying three new servers in every single region just to serve as a Consul staging environment. Just too much expense at this stage for the company. So if I'm going to make a staging environment globally, I need them to run in parallel. My first attempt at hacking on the production Consul instances resulted in my pushing bad configuration and taking down one of our regions in production. ([Here's the StatusPage report!](https://livepeer.statuspage.io/incidents/lk8g3g2s9rn5)). 

2. I need to provide continuity of service at the same HTTPS addresses for the old local Consul system and the new WAN-federated Consul system while we're working on the changeover. So I was planning on pointing Traefik at both the old Consul servers and the new Consul servers, which would also require them to run in parallel. Another, related use case would be doing some manner of blue/green deploy of a new Consul version in parallel with the old one.

So yeah. Any ideas on how I could accomplish such a thing without the ability to run side-by-side Consul Releases?

**Comment by david-yu** [comment]
Hi @iameli, thank you for the incredible detail and its good to hear you are finding Consul working well at LivePeer. Unfortunately as @ndhanushkodi had mentioned that the use case for running multiple Consul clusters on a single set of servers is incredibly hard for us to support in Consul. You can definitely continue to work on customizing your own Helm chart to support this use case, however, it is not a standard deployment model that we currently support and won't likely support in the future. 

I did want to mention that there is an enterprise feature called [Namespaces](https://www.consul.io/docs/enterprise/namespaces) for KV that can help with the particular use case you are looking to accomplish as you can see with the API: https://www.consul.io/api-docs/kv. It is an enterprise feature so that may not be what you're looking for. Determining whether a a feature is enterprise vs not is something we are conscious about determining so that it helps when managing larger more complex deployments of Consul at an organizational scale. 

Sorry if that is not the answer you are looking for, but we hope that gives you some clarity on the direction of the architecture.

**Comment by david-yu** [comment]
Hi @iameli I will close this issue as we don't intend to support more than one release of Consul per cluster as part of our roadmap as it is incredibly hard to support such a complex environment. There is another PR you can track which is #999 which is exploring the possibility of customizing client ports.
